man/ply.1.ronn
ply(1) -- dynamically instrument the kernel
===========================================

## SYNOPSIS

`ply` <program-file> <br>
`ply` -c <program-text>

## DESCRIPTION

ply dynamically instruments the running kernel to aggregate and
extract user-defined data. It compiles an input program to one or more
Linux bpf(2) binaries and attaches them to arbitrary points in the kernel
using kprobes and tracepoints.

## OPTIONS

  * `-A`, `--ascii`:
    Restrict output to ASCII, no Unicode runes.

  * `-c`, `--command`:
    The program is supplied as an argument, rather than in a file.

  * `-d`, `--debug`:
    Enable debugging output.

  * `-D`, `--dump`:
    Do not execute the program, instead dump the generated Linux BPF
    instructions.

  * `-h`, `--help`:
    Print usage message.

  * `-t`, `--timeout`=<seconds>:
    Terminate the program after the specified time.

  * `-u`, `--usekern`;
    Use the current kernel version for the constructed eBPF program, rather than the
    version that is compiled into the program. Potentially this introduces problems,
    it is better to recompile the program to match the current kernel.

  * `-v`, `--version`:
    Print version information.


## SYNTAX

The syntax is C-like in general, taking its inspiration dtrace(1) and,
by extension, from awk(1).


### Probes

A program consists of one or more _probes_, which are analogous to
awk's pattern-action statements. The syntax for a probe is as follows:

    provider:probe-definition ['/' predicate '/']
    {
         statement ';'
        [statement ';' ... ]
    }

The _provider_ selects which probe interface to use. See the PROVIDERS
section for more information about each provider. It is then up to the
provider to parse the _probe-definition_ to determine the point(s) of
instrumentation.

When tracing, it is often desirable to filter events to match some
criteria. Because of this, ply allows you to provide a _predicate_,
i.e. an expression that must evaluate to a non-zero value in order for
the probe to be executed.

Then follows a block of _statements_ that perform the actual information
gathering. All but the last statement of a probe must be terminated
with a semi-colon. Specifically, a simple probe containing only one
statement does not need one.

A provider may define a default probe to be used if the user does not
supply a block.


### Control of Flow

Probes support two different types of flow control, the first is the
_if-statement_, which conforms to the same rules as C's equivalent:

    'if' '(' expr ')'
        statement ';' | block
    [else
        statement ';' | block]

In order to ensure that a probe will have a finite run-time the kernel
does not allow backwards branching. As a result, ply does not have a
traditional loop construct like _for_ or _while_. Instead it provides
an _unroll_ statement that simply copies the contained _statements_
_N_ times to the output program:

    'unroll' '(' N ')'
        statement ';' | block


### Type System

The type system is rudimentary, containing only three (3) types:

  * `Numbers`:
    Represented as 64-bit signed integers, initialized to zero (0).

  * `Strings`:
    Of arbitrary length, initialized to the empty string.

  * `Records`:
    Compound type holding any number of `Numbers` and `Strings` with
    each component initialized to its zero value.

Numbers and string literals are specified in the same way as in
C. Records use the following syntax:

    '[' expr [ ',' expr ... ] ']'


### Maps

The primary way to extract information is to store it in a _map_,
i.e. in a hash table. Like awk(1), ply dynamically creates any
referenced maps and their key and value types are inferred from the
context in which they are used. All maps are in the global scope and
can thus be used both for extracting data to the end-user, and for
carrying data between probes. A map name must begin with the at sign
'@' and then follows the rules of identifiers from C. For small
scripts where only one map is used, it is convenient and idiomatic to
simply name it '@'. After the name comes a record specifier which is
used as the key in the map.

    @mapname[exprs]

Data can be stored in a map by assigning a value to a given key:

    @mapname[exprs] = expr

If a map key is assigned the special value _nil_, the key is deleted
and will return its zero value if referenced again.

A more common way is to _aggregate_ data using _methods_, i.e. functions
that operate on the data stored in a map:

  * `@mapname[exprs].count()`:
    Bumps a counter.

  * `@mapname[exprs].quantize(number-expr)`:
    Evaluates the argument and aggregates on the most significant bit
    of the result. In other words, it stores the distribution of the
    expression.


### Variables

Variables are typically used for storing temporary data, like loop
counters or results of computations that are used multiple times in a
probe. They are local to the probe in which they are defined, i.e. the
same identifier may be used independently in multiple probes. Their
names follows the C identifier rules.


## PROVIDERS

A _provider_ makes data available to the user by exporting functions
to the probe. Function calls use the same syntax as most languages
that inherit from C. In addition to the provider-specific functions,
all providers inherits a set of common functions:

  * `comm()`, `execname()` => string:
    Returns the _name_ of the running process's executable.

  * `cpu()` => number:
    Returns the _CPU ID_ of the processor on which the probe fired.

  * `gid()` => number:
    Returns the _group ID_ of the running process.

  * `log2(number-expression)` => number:
    Returns the logarithm, base 2, of the argument.

  * `mem(address, format)` => TYPE:

    Copy from _address_, using the _format_ specifier to determine the
    desired size and type. The format specifier takes inspiration from
    python(1)'s _struct_ module:

    [repeat]type[[<repeat>]<type> ...]

    _repeat_ is an optional number, indicating that this member is an
    array of the following type. The recognized types are:

    _b_/_B_: Signed/unsigned  8-bit integer. <br>
    _h_/_H_: Signed/unsigned 16-bit integer. <br>
    _w_/_W_: Signed/unsigned 32-bit integer. <br>
    _q_/_Q_: Signed/unsigned 64-bit integer. <br>
    _i_/_I_: Signed/unsigned    int integer. <br>
    _l_/_L_: Signed/unsigned   long integer. <br>
    _s_: String. _repeat_ means maximum length. <br>
    _p_: Pointer.

    If more than one type is specified, _mem()_'s output will be
    of record type.

  * `nsecs()` => number:
    Returns the time since the system started, in nanoseconds.

  * `pid()` => number:
    Returns the _process ID_ of the running process. Also known as
    _thread group ID_ by the kernel.

  * `printf(format [, expression, expression ... ])`:
    Prints _formatted output_ to ply's standard out. In addition to
    the formats recognized by the printf sitting in your <stdio.h>,
    ply's also recognizes '%v' which will dump the value according to
    the inferred type's default.

    Beware that while there are times when it is useful to print data
    from a probe, it is very often not the best way of obtaining the
    insight that is sought.

    Many of the kernel's functions may be called at rates well above
    the 1k calls/s mark, at which point making sense of print output
    becomes very hard. In such cases, consider using some type of
    aggregation to limit the amount of data. Then, once you have
    zeroed in on the problem, printing might become useful.

  * `secs()` => number:
    Returns the time since the system started, in seconds.

  * `sizeof(format)` => number:
    Returns the number of bytes required to store an object specified
    in _format_. See _mem()_ for the interpretation of the format
    specifier.

  * `strcmp(string-expression, string-expression)` => number:
    Returns -1, 0 or 1 if the first argument is less than, equal to or
    greater than the second argument respectively. Strings are
    compared by their lexicographical order.

  * `tid()` => number:
    Returns the _task ID_ of the running process, i.e. the kernel's
    identifier for the current thread of execution.

  * `uid()` => number:
    Returns the _user ID_ of the running process.


### kprobe and kretprobe

These providers use the corresponding kernel features to instrument
arbitrary instructions in the kernel. The _probe-definition_ may be
either an address or a symbol name. When using a symbol name, glob
expansion is performed allowing a single probe to be inserted at
multiple locations. E.g. _kprobe:SyS\_*_, would match every syscall.

Shared functions:

  * `reg(number)`, `reg(string)` => number:
    If called with a number, it returns the value of the n:th CPU
    _register_, according to the order in the architecture specific
    _pt_regs_ struct. If called with a string, the string is looked up
    against the register names as they are defined in the _pt_regs_
    struct.

  * `stack()` => number:
    Get a _stack trace_ of the current probe. This is just returns an
    index into a separate map containing the actual instruction
    pointers. As a user though, you can think of this function as
    returning a string containing the stack trace at the current
    location. Indeed _printf("%v\n", stack())_ will produce exactly
    that.

_kprobe_ specific functions:

  * `arg(number)` => number:
    Returns the value of the specified _argument_ of the function to
    which the probe was attached, zero-indexed. I.e. arg(0) is the 1st
    argument, arg(1) is the 2nd, and so on.

    CAUTION: This is just syntactic sugar on top of _reg()_, using the
    relevant architecture's calling convention as a mapping. If your
    compiler decides to optimize out arguments or do other sneaky
    things, ply will be utterly oblivious to that.

  * `func()`, `probefunc()` => number:
    Returns the kernel address of the _function_ to which the probe
    was attached. It also marks it as being a symbol, which means that
    _printf()_ will know to print it as a symbol if possible.

    This is also just syntactic sugar on top of _reg()_ mapping to the
    architecture's instruction pointer register.

_kretprobe_ specific function:

  * `retval()` => number:
    Returns the return value of the probed function.


### trace

The _trace_ provider allows you to attach probes to the kernel's
_tracepoint_ infrastructure. The _probe-definition_ is matched against
the _debugfs_ path located at
_/sys/kernel/debug/tracing/events_. E.g. _trace:sched/sched_switch_
would run the probe every time the kernel schedules a new task.

A tracepoint's data is defined in its _format_ file,
e.g. _/sys/kernel/debug/tracing/events/sched/sched_switch/format_. The
trace provider will export each field as a function. Say the _format_
file has the following content:

    name: sched_switch
    ID: 287
    format:
	field:unsigned short common_type;	offset:0;	size:2;	signed:0;
	field:unsigned char common_flags;	offset:2;	size:1;	signed:0;
	field:unsigned char common_preempt_count;	offset:3;	size:1;	signed:0;
	field:int common_pid;	offset:4;	size:4;	signed:1;

	field:char prev_comm[16];	offset:8;	size:16;	signed:1;
	field:pid_t prev_pid;	offset:24;	size:4;	signed:1;
	field:int prev_prio;	offset:28;	size:4;	signed:1;
	field:long prev_state;	offset:32;	size:8;	signed:1;
	field:char next_comm[16];	offset:40;	size:16;	signed:1;
	field:pid_t next_pid;	offset:56;	size:4;	signed:1;
	field:int next_prio;	offset:60;	size:4;	signed:1;

The probe may then call the following functions:

  `common_type()` => number <br>
  `common_flags()` => number <br>
  `common_preempt_count()` => number <br>
  `common_pid()` => number

  `prev_comm()` => string <br>
  `prev_pid()` => number <br>
  `prev_prio()` => number <br>
  `prev_state()` => number <br>
  `next_comm()` => string <br>
  `next_pid()` => number <br>
  `next_prio()` => number <br>


## EXAMPLE

### Extracting data

Print all opened files on the system, and who opened them:

    kprobe:SyS_open
    {
        printf("%v: %v\n", comm(), mem(arg(0), "128s"))
    }

### Quantize

Record the distribution of the return value of read(2):

    kretprobe:SyS_read
    {
        @.quantize(retval())
    }


### Wildcards

Count all syscalls made on the system, grouped by function:

    kprobe:SyS_*
    {
        @[func()].count()
    }


Count all syscalls made by every dd(1) process, grouped by function:

    kprobe:SyS_* / !strcmp(execname, "dd") /
    {
        @[func()].count()
    }


### Object Tracking

Record the distribution of the time it takes an _skb_ to go from
_netif_receive_ to _ip_rcv_:

    kprobe:__netif_receive_skb_core
    {
        @rx[arg(0)] = nsecs()
    }

    kprobe:ip_rcv / @rx[arg(0)] /
    {
        @diff.quantize(nsecs() - @rx[arg(0)]);
        @rx[arg(0)] = nil;
    }


## RETURN VALUE

  * `0`:
    Program was successfully compiled and loaded into the kernel.

  * `Non-Zero`:
    An error occurred during compilation or during kernel setup.


## AUTHORS

Tobias Waldekranz <tobias@waldekranz.com>

## COPYRIGHT

Copyright 2015-2017 Tobias Waldekranz

License: GPLv2


## SEE ALSO

awk(1) dtrace(1) bpf(2)


src/ply.c
/*
 * Copyright 2015-2017 Tobias Waldekranz <tobias@waldekranz.com>
 *
 * This file is part of ply.
 *
 * ply is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, under the terms of version 2 of the
 * License.
 *
 * ply is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ply.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <ctype.h>
#include <getopt.h>
#include <linux/version.h>
#include <signal.h>
#include <stdio.h>
#include <sys/resource.h>
#include <unistd.h>

#include <ply/ast.h>
#include <ply/evpipe.h>
#include <ply/map.h>
#include <ply/ply.h>
#include <ply/pvdr.h>

#include "config.h"

FILE *scriptfp;

struct globals G= {
	.kernel_version = LINUX_VERSION_CODE,
};

static const char *sopts = "AcdDht:uv";
static struct option lopts[] = {
	{ "ascii",   no_argument,       0, 'A' },
	{ "command", no_argument,       0, 'c' },
	{ "debug",   no_argument,       0, 'd' },
	{ "dump",    no_argument,       0, 'D' },
	{ "help",    no_argument,       0, 'h' },
	{ "timeout", required_argument, 0, 't' },
	{ "usekern", no_argument,       0, 'u' },
	{ "version", no_argument,       0, 'v' },

	{ NULL }
};

static void getkernelversion()
{
	FILE *f = fopen("/proc/sys/kernel/version", "r");
	char buf[128];
	if (!f)
		return;
	fgets(buf, sizeof(buf), f);
	fclose(f);
	/* Parse it by hand, no need to pull out a regular expression */
	for (const char *s = buf; *s; s++) {
		char *p1;
		int patch, minor;
		if (!isdigit(s[0]))
			continue;	/* Single Digit */
		if (s[1] != '.')
			continue;
		if (!isdigit(s[2]))
			continue;
		patch = strtol(s + 2, &p1, 10) & 255;
		if (p1[0] != '.')
			continue;
		if (!isdigit(p1[1]))
			continue;
		minor = strtol(p1 + 1, NULL, 10) & 255;
		G.kernel_version = ((*s - '0') << 16 + (patch << 8) + minor);
		return;
	}
}


static void usage()
{
	puts("ply - Dynamic tracing utility\n"
	     "\n"
	     "Usage:\n"
	     "  ply [options] <script_file>\n"
	     "  ply [options] -c <script_string>\n"
	     "\n"
	     "Options:\n"
	     "  -A                  ASCII output only, no Unicode.\n"
	     "  -c <script_string>  Execute script literate.\n"
	     "  -d                  Enable debug output.\n"
	     "  -D                  Dump generated BPF and exit.\n"
	     "  -h                  Print usage message and exit.\n"
	     "  -t <timeout>        Terminate trace after <timeout> seconds.\n"
	     "  -u                  Use kernel value of LINUX_VERSION_CODE.\n"
	     "  -v                  Print version information.\n"
		);
}

static void version()
{
	fputs(PACKAGE "-" VERSION, stdout);
	if (strcmp(VERSION, GIT_VERSION))
		fputs("(" GIT_VERSION ")", stdout);

	printf(" (linux-version:%u~%u.%u.%u)\n",
	       G.kernel_version,
	       (G.kernel_version >> 16) & 0xff,
	       (G.kernel_version >>  8) & 0xff,
	       (G.kernel_version >>  0) & 0xff);
}

static int parse_opts(int argc, char **argv, FILE **sfp)
{
	int cmd = 0;
	int opt;

	G.map_nelem = 0x400;

	while ((opt = getopt_long(argc, argv, sopts, lopts, NULL)) > 0) {
		switch (opt) {
		case 'A':
			G.ascii = 1;
			break;
		case 'c':
			cmd = 1;
			break;
		case 'd':
			G.debug = 1;
			break;
		case 'D':
			G.dump = 1;
			break;
		case 'h':
			usage(); exit(0);
			break;
		case 't':
			G.timeout = strtol(optarg, NULL, 0);
			if (G.timeout <= 0) {
				_e("timeout must be a positive integer");
				usage(); exit(1);
			}
			break;
		case 'u':
			getkernelversion();
			break;
		case 'v':
			version(); exit(0);
			break;

		default:
			_e("unknown option '%c'", opt);
			usage(); exit(1);
			break;
		}
	}

	if (cmd)
		*sfp = fmemopen(argv[optind], strlen(argv[optind]), "r");
	else if (optind < argc)
		*sfp = fopen(argv[optind], "r");
	else {
		_e("no input");
		usage(); exit(1);
	}

	if (!*sfp) {
		_eno("unable to read script");
		usage(); exit(1);
	}

	return 0;
}

static void memlock_uncap(void)
{
	struct rlimit limit;
	rlim_t current;
	int err;

	err = getrlimit(RLIMIT_MEMLOCK, &limit);
	if (err) {
		_eno("unable to retrieve memlock limit, "
		     "maps are likely limited in size");
		return;
	}

	current = limit.rlim_cur;

	/* The total size of all maps that ply is allowed to create is
	 * limited by the amount of memory that can be locked into
	 * RAM. By default, this limit can be quite low (64kB on the
	 * author's system). So this simply tells the kernel to allow
	 * ply to use as much as it needs. */
	limit.rlim_cur = limit.rlim_max = RLIM_INFINITY;
	err = setrlimit(RLIMIT_MEMLOCK, &limit);
	if (err) {
		const char *suffix = "B";

		if (!(current & 0xfffff)) {
			suffix = "MB";
			current >>= 20;
		} else if (!(current & 0x3ff)) {
			suffix = "kB";
			current >>= 10;
		}

		_eno("could not remove memlock size restriction");
		_w("total map size is limited to %lu%s", current, suffix);
		return;
	}

	_d("unlimited memlock");
}

static int term_sig = 0;
static void term(int sig)
{
	term_sig = sig;
	return;
}

int main(int argc, char **argv)
{
	evpipe_t *evp;
	node_t *probe, *script;
	prog_t *prog = NULL;
	pvdr_t *pvdr;
	FILE *sfp, *enable;
	int err = 0, num, total;

	scriptfp = stdin;
	err = parse_opts(argc, argv, &sfp);
	if (err)
		goto err;

	G.ksyms = ksyms_new();
	memlock_uncap();

	script = node_script_parse(sfp);
	if (!script) {
		err = -EINVAL;
		goto err;
	}

	err = pvdr_resolve(script);
	if (err)
		goto err;

	err = annotate_script(script);
	if (err)
		goto err;

	evp = calloc(1, sizeof(*evp));
	assert(evp);
	script->dyn->script.evp = evp;

	err = evpipe_init(evp, 4 << 10);
	if (err)
		goto err;

	err = map_setup(script);
	if (err)
		goto err;
		
	if (G.dump)
		node_ast_dump(script);

	total = 0;
	node_foreach(probe, script->script.probes) {
		err = -EINVAL;
		prog = compile_probe(probe);
		if (!prog)
			break;

		if (G.dump)
			continue;

		pvdr = node_get_pvdr(probe);
		num = pvdr->setup(probe, prog);
		if (num < 0)
			break;

		total += num;
	}

	if (G.dump)
		goto done;
	
	if (num < 0) {
		err = num;
		goto err;
	}

	if (G.timeout) {
		siginterrupt(SIGALRM, 1);
		signal(SIGALRM, term);
		alarm(G.timeout);
	}

	siginterrupt(SIGINT, 1);
	signal(SIGINT, term);
	
	enable = fopen("/sys/kernel/debug/tracing/events/enable", "w");
	if (!enable) {
		perror("unable to enable probes");
		err = -errno;
		goto err;
	}

	fputs("1\n", enable);
	fflush(enable);
	rewind(enable);

	fprintf(stderr, "%d probe%s active\n", total, (total == 1) ? "" : "s");
	err = evpipe_loop(evp, &term_sig, 0);

	fprintf(stderr, "de-activating probes\n");
	fputs("0\n", enable);
	fflush(enable);
	fclose(enable);

	map_teardown(script);

	node_foreach(probe, script->script.probes) {
		pvdr = node_get_pvdr(probe);
		err = pvdr->teardown(probe);
		if (err)
			break;
	}

	/* disable all kprobe events */
	fclose(fopen("/sys/kernel/debug/tracing/kprobe_events", "w"));
done:
err:
	if (prog)
		free(prog);
	if (script)
		node_free(script);

	return err ? 1 : 0;
}
